### DISCUSSION TITLE
Predicable object iteration order

### DISCUSSION THREAD

**Discussion by gibson042** [jmespath-community/jmespath.spec#94] [Ideas]
Spun off from https://github.com/jmespath-community/jmespath.spec/discussions/47#discussioncomment-3317114

JavaScript learned the hard way that "_implementations are not required to return values in any specific order_" is not really a tenable position for a language with multiple implementations, because in practice if a sufficiently popular implementation does provide results in a predictable order then users come to rely upon that behavior (cf. [Hyrum's Law](https://www.hyrumslaw.com/)) and less popular implementations are forced to follow suit, resulting in a de facto predictable order that eventually gets added to the specification anyway. I am only aware of two ways around this: either specify a predictable order up front, or specify that order is intended to be unpredictable in any particular implementation (such that an implementation which sorted entries or preserved source order would be nonconforming, which is basically impossible to ensure with conformance tests).

I absolutely understand the desire to let the use of predictable vs. unpredictable ordering be unspecified and therefore support maximum convenience for building an implementation, but my experience is that it just doesn't work out in the long run.

Accordingly, `keys` and `values` and a potential [`items`](https://github.com/jmespath-community/jmespath.spec/discussions/47) should all be specified to return results in a particular order. Using source order is possible and would be acceptable, but I'm more partial to specifying a static algorithm by which keys must be sorted (e.g., lexicographically by Unicode code point) because that minimizes potential security vulnerabilities derived from leveraging information about insertion order.

**Comment by springcomp** [comment]
Thanks a lot for that.

I'm completely aware that Hyrum's Law will kick in. That's why I suggested [standardizing an undocumented behaviour](https://github.com/jmespath-community/jmespath.spec/discussions/82).

By any chance, would you feel like drafting a JEP here ?

**Reply by gibson042** [reply]
> By any chance, would you feel like drafting a JEP here ?

I'm happy to draft a JEP, but not sure of next steps. I opened this discussion by following https://github.com/jmespath-community/jmespath.spec#guidelines-for-proposing-new-features :
> Then open a [discussion](https://github.com/jmespath-community/jmespath.spec/discussions) to discuss the feature, its merit, and any possible alternatives. Once the discussion has reached a mature level of feedback it will be assigned a JEP#. At this point it is safe to begin composing a JEP, along with all the required changes and documentation.

**Reply by springcomp** [reply]
Yes, that’s a good point we should clarify. To be honest, this is the first community open-source project I’m trying to manage.

The idea is that discussions are a good first point to gauge whether changes in the spec would be warranted. In [all the cases](https://github.com/jmespath-community/jmespath.spec/discussions/38) that you’ve seen so far, we have been having those discussions @innovate-invent and I almost exclusively, which allowed us to reach a consensus fairly quickly. That’s why I drafted [a set of JEPs](https://github.com/jmespath-community/jmespath.spec/discussions/97) that I’m pretty confident can be accepted by a wide proportion of community users.

It’s a good thing that other members of the community such as yourselves get involved.

For the sake of complete transparency and honesty, I must stress that [I started this community effort](https://github.com/jmespath/jmespath.site/issues/94) out of frustration that @jamesls was not able to dedicate as much time and provide as much feedback as I feel would be warranted to stewart changes to the specification.

The goal has always been to foster proposals and help improve JMESPath. However, as it seems progress on the official JMESPath specification is effectively stalled, I decided to go ahead and pursue this community effort. Our objective is to implement those improvements and push them as pull-requests to interested parties – most notably, the [AWS CLI](https://github.com/aws/aws-cli).

That said, I commit to [share those changes](https://github.com/jmespath/jmespath.site/issues/94#issuecomment-1136380999) upstream, as [I have done recently](https://github.com/jmespath/jmespath.site/issues/94#issuecomment-1136778209), anytime @jamesls will provide feedback and guidance.

So for the purpose of this community effort, if you feel like investing some time drafting a JEP here in the discussions, you will get feedback to refine it until you or someone can propose a pull request to the [JMESPath Community](https://jmespath.site) website.

**Reply by springcomp** [reply]
Might need to incorporate part of the[ JSON Canonicalization Scheme](https://www.rfc-editor.org/rfc/rfc8785).

_This document describes the JSON Canonicalization Scheme (JCS). This specification defines how to create a canonical representation of JSON [[RFC8259](https://www.rfc-editor.org/rfc/rfc8785#RFC8259)] data by building on the strict serialization methods for JSON primitives defined by ECMAScript [[ECMA-262](https://www.rfc-editor.org/rfc/rfc8785#ECMA-262)], constraining JSON data to the I-JSON [[RFC7493](https://www.rfc-editor.org/rfc/rfc8785#RFC7493)] subset, and by **using deterministic property sorting**._

**Reply by gibson042** [reply]
That's not sufficient, because JMESPath is not constrained to I-JSON.

**Reply by innovate-invent** [reply]
The more I think about this, the more I think that this should be left for the application to define. Forcing an order could be problematic to the underlying data structure types.

**Reply by springcomp** [reply]
> That's not sufficient, because JMESPath is not constrained to I-JSON.

@gibson042  I’m not sure I follow. I was referring to the JSON that JMESPath processes as input and produces as output.
To me, while we cannot make any guarantee from the JSON received as input, I think we can mandate that the output produced by JMESPath does indeed comply to the I-JSON specification.

My point, however, was not to actually mandate that.

Rather, maybe we can take the referred to JCS spec as an inspiration to provide a compatible ordering that does not conflict.
Specifically, if we were to choose a deterministic order for object properties, maybe following [guidelines set forth](https://www.rfc-editor.org/rfc/rfc8785#name-sorting-of-object-propertie) in the spec would be a good idea.

**Reply by gibson042** [reply]
> I think we can mandate that the output produced by JMESPath does indeed comply to the I-JSON specification.

It would be possible, but as a separate change and backwards-incompatible (e.g., observe at https://jmespath.site/ how the non-I-JSON ``` `"\uD800\uDEAD"` ``` is round-trippable).

> Rather, maybe we can take the referred to JCS spec as an inspiration to provide a compatible ordering that does not conflict.

:+1:

**Reply by gibson042** [reply]
> The more I think about this, the more I think that this should be left for the application to define. Forcing an order could be problematic to the underlying data structure types.

Understandable, but what mechanism would prevent the ecosystem from de facto requiring consistency with a predictable ordering exhibited by popular implementations? If the prevailing convention is e.g. source order and people start to rely upon that, then bugs will be filed against implementations with unpredictable ordering and they'll capitulate despite technically conforming with the spec. Implementation freedom with respect to observable behavior is a hazard in general, but for this case in particular there is already precedent that can hopefully be learned from (cf. https://github.com/tc39/proposal-for-in-order ).

**Reply by springcomp** [reply]
> … (e.g., observe at https://jmespath.site/ how the non-I-JSON `` `"\uD800\uDEAD"` `` is round-trippable).

I would be interested to understand more about this observation.

Seems to me, that `"\uD800\uDEAD"` is a valid I-JSON encoding of an [otherwise invalid Unicode string](https://www.lookout.net/2011/06/special-unicode-characters-for-error.html) (it is explicitly called as valid in section 2.1 of that spec).

What do you mean by round-trippable ? That it can be converted to an in-memory representation of a string and serialized back to the same JSON escaped string ?

**Reply by springcomp** [reply]
> The more I think about this, the more I think that this should be left for the application to define. Forcing an order could be problematic to the underlying data structure types.

@innovate-invent I understand your point and would tend to agree when requiring a specific sort order. 
Would you still feel that way if we merely required that _source order be preserved_ ?

**Reply by innovate-invent** [reply]
That depends on the definition of "source order". I generally use JMESPath to manipulate arbitrary data structures that do not necessarily originate from a document, or even JSON. For example, a 'object' can be represented (in Python) by a dict, a arbitrary class type, or any other associative data type. A 'array' can be represented as a native list, generator, or any class type with the `__getitem__` function accepting a number. These arbitrary types can have their own defined order that they iterate over their sub-elements. If the underlying object is the source, then "source order" would be acceptable, otherwise it really needs to be left to the application to define.

It is easy to forget that JMESPath is a library, and certain allowances need to be made to not over-constrain the importing application.

**Reply by gibson042** [reply]
> > … (e.g., observe at https://jmespath.site/ how the non-I-JSON `` `"\uD800\uDEAD"` `` is round-trippable).
> 
> I would be interested to understand more about this observation.
> 
> Seems to me, that `"\uD800\uDEAD"` is a valid I-JSON encoding of an [otherwise invalid Unicode string](https://www.lookout.net/2011/06/special-unicode-characters-for-error.html) (it is explicitly called as valid in section 2.1 of that spec).

I apologize for the mistake. `"\uD800\uDEAD"` is absolutely valid (as is its unescaped analog, `"𐊭"`). I meant to use `"\uDEAD"`, which _does_ encode an unpaired surrogate and therefore is not valid I-JSON.

> What do you mean by round-trippable ? That it can be converted to an in-memory representation of a string and serialized back to the same JSON escaped string ?

Basically, yes. Not necessarily the same JSON string, but one that represents exactly the same data (in this case, `"\uDEAD"` → `"\udead"`) and is therefore treated equivalently if reused as input.
